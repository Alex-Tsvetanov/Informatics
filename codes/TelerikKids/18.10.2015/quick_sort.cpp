#include <iostream>
#include <random>

using namespace std;

template <typename T>
T* get_pivot (T* a, T* b)
{
	T* masterPivot = a;
	T averrage = 0;
	for (T* it = a ; it != b ; it ++)
		averrage += (*it);
	averrage /= (b - a);
	for (T* it = a ; it != b ; it ++)
		if (abs ((*masterPivot) - averrage) > abs ((*it) - averrage))
			masterPivot = it;

	return masterPivot;
}

template <typename T>
T* fillArr (T* newArr, T* pivot, T* s, T* e)
{
	T* indOfPivot;
	size_t ind = 0;
	for (T* it = s ; it != e ; it ++)
		if ((*it) < (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	indOfPivot = newArr + ind;
	for (T* it = s ; it != e ; it ++)
		if ((*it) >= (*pivot))
		{
			newArr [ind] = *it;
			ind ++;
		}
	return indOfPivot;
}

template <typename T>
void copyArr (T* newArr, T* s, T* e)
{
	T* newArrIt = newArr;
	for (T* it = s ; it != e ; it ++, newArrIt ++)
		*it = *newArrIt;
}

template <typename T>
void quick_sort (T* s, T* e)
{
	if (s + 1 == e)
		return ;
	if (s == e)
		return ;
	T* newArr = new T [e - s];
	T* pivot = fillArr (newArr, get_pivot (s, e), s, e);
	//for (T* it = s ; it != e ; it ++)
	//	cout << *it << " ";
	//cout << "\n" << *pivot << "\n----------------------------" << endl;
	quick_sort (newArr, pivot + 1);
	quick_sort (pivot + 1, newArr + (e - s));
	copyArr (newArr, s, e);
	delete[] newArr;
}

int main ()
{
	size_t SIZE;
	cin >> SIZE;
	int* arr = new int [SIZE];
	for (size_t i = 0 ; i < SIZE ; i ++)
		cin >> arr [i];
	quick_sort (arr + 0, arr + SIZE);
	for (size_t i = 0 ; i < SIZE ; i ++)
		cout << arr [i] << " ";
	cout << "\n";
}
