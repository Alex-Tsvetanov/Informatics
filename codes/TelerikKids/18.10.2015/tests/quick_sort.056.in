62
79 41 23 54 36 73 93 61 90 89 63 46 92 58 66 64 87 2 8 25 65 59 18 18 29 73 54 1 30 3 73 77 96 73 55 39 75 75 26 95 41 56 3 93 32 47 25 94 3 18 50 90 13 86 57 24 21 71 24 62 40 96 
