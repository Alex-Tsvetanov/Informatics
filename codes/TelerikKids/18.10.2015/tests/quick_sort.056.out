1 2 3 3 3 8 13 18 18 18 21 23 24 24 25 25 26 29 30 32 36 39 40 41 41 46 47 50 54 54 55 56 57 58 59 61 62 63 64 65 66 71 73 73 73 73 75 75 77 79 86 87 89 90 90 92 93 93 94 95 96 96 
